<!DOCTYPE html>
<!--
~ RED: RNA Editing Detector
~     Copyright (C) <2014>  <Xing Li>
~
~     RED is free software: you can redistribute it and/or modify
~     it under the terms of the GNU General Public License as published by
~     the Free Software Foundation, either version 3 of the License, or
~     (at your option) any later version.
~
~     RED is distributed in the hope that it will be useful,
~     but WITHOUT ANY WARRANTY; without even the implied warranty of
~     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
~     GNU General Public License for more details.
~
~     You should have received a copy of the GNU General Public License
~     along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->

<html>
<head>
    <title>What is RED</title>
    <style type="text/css">
        body {
        font-family: Times New Roman;
        }
        p,ul {
        font-size: 18px;
        }
        li {
        padding-bottom: 1 em
        }
        .imgname {
        text-align: center;
        font-size: 10px;
        font-style: italic;
        }
    </style>
</head>
<body>
<h3 id="section1.1">1.1 What is RED</h3>

<p>RNA editing is one of the post- or co-transcriptional processes with modification of RNA nucleotides from their genome-encoded sequence. In human, RNA
    editing event occurs mostly by deamination of adenosine to inosine (A-to-I) conversion through ADAR enzymes.</p>

<p>RED (RNA Editing Detector) is a program to detect and visualize RNA editing events at the genomic scale using next-generation sequencing (NGS) data. Th
    e software integrates multiple steps to filter spurious variant sites, and then performs formal statistical tests to identify reliable RNA editing
    sites. A potential RNA editing site can be visualized in the context of reference genome (/gene).</p>

<div align="center"><img src="../img/fig.1.1.png"/></div>
<p class="imgname">Fig. 1.1 A screen shot of RED</p>

<p>RED is a Java-based program with Graphic User Interface (GUI), which works on mainstream operating systems (e.g., Windows, Linux and Mac OS X). To make
    it more efficient, everlasting and faster-saving, we use MySQL database management system (DBMS) in manipulating huge NGS data. In addition, we use
    functions in <em>R</em> to perform statistical analysis. However, you'll never manipulate R and MySQL when you use RED. All of the operation relative to
    <em>R</em> and MySQL is completed by RED automatically. RED enables the analysis of RNA editing events at two levels (illustrated below) using VCF files
    from different sources and the presentation of full mapped sequence data using BAM/SAM file:</p>
<ul>
    <li>
        Detection of RNA editing sites using RNA-seq data alone (i.e., <em>De Novo</em> identification)
    </li>
    <li>
        Detection of RNA editing sites using both DNA-seq data (e.g., whole exome sequencing) and RNA-Seq data
    </li>
</ul>

</body>
</html>